Add genotype matrix script.

BonfaceKilz opened the pull request.

BonfaceKilz committed:
Add genotype matrix script.

* scripts/lmdb_matrix.py: New file.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Simplify genotype db script.

Hashing and versioning made the script very slow.  Now, this script
can handle much larger data.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

Alexanderlacuna commented on scripts/lmdb_matrix.py:
```diff
@@ -0,0 +1,164 @@
+    metadata = json.dumps(genotype.metadata).encode("utf-8")
+    with db.begin(write=True) as txn:
+        txn.put(b"matrix", genotype.matrix.tobytes())
+        txn.put(b"transpose", genotype.transpose.tobytes())
```
Question: Why do you need to store the  transposed matrix?

BonfaceKilz replied:
I've removed this in my updated work.  You are right, we don't need this for now; since that adds un-necessary overhead for a really huge matrix.

Alexanderlacuna commented on scripts/lmdb_matrix.py:
```diff
@@ -0,0 +1,164 @@
+        # Read header
+        header = stream.readline().strip().split()
+        metadata_columns = ["Chr", "Locus", "cM",
+                            "Mb"] if "Mb" in header else ["Chr", "Locus", "cM"]
```
suggestion:  why not use a unified structure for all files with all the provided headers and using a value such as 'N/A' when a header value is missing?

Alexanderlacuna commented on scripts/lmdb_matrix.py:
```diff
@@ -0,0 +1,164 @@
+        for i in range(nrows):
+            line = stream.readline().strip().split()
+            meta, data = line[:len(metadata_columns)
+                              ], line[len(metadata_columns):]
```
Suggestion: Could you add a comment for this just to make it clear what you are fetching?
i.e # this returns chr ... etc

Alexanderlacuna commented on scripts/lmdb_matrix.py:
```diff
@@ -0,0 +1,164 @@
+            meta, data = line[:len(metadata_columns)
+                              ], line[len(metadata_columns):]
+            for j, element in enumerate(data):
+                if element.isdigit():
```
Question:  
Since the objective of this is to obtain the chromosome value, what happens with the sex chromosomes which is a string, for example, the 'X' chromosome ?

Alexanderlacuna commented on scripts/lmdb_matrix.py:
```diff
@@ -0,0 +1,164 @@
+                    matrix[i, j] = int(element)
+                elif element == maternal:
+                    matrix[i, j] = 0
+                elif element == paternal:
```
suggestion: I think this if else block to get  geno_codes is too rigid,
how about something like:
```py
  genotype_map = {
    maternal: 0,
    paternal: 1,
    heterozygous: 2,
    unknown: 3
}
value = value = genotype_map.get(element)
matrix[i, j ] =value
```
With this you can extend the number of genotype categories, you can also  add that as a part of the metadata

BonfaceKilz replied:
But they are that rigid lolz.  Regardless, `match ... case` FTW.

Alexanderlacuna commented:
@BonfaceKilz  is this task complete?

BonfaceKilz committed:
Improve genotype file parsing.

* scripts/lmdb_matrix.py (GenotypeMatrix): Delete transpose.  Add
"file_info" field.
(save_excursion, count_lines): Delete.
(get_genotype_dimensions, read_genotype_headers): New function.
(read_geno_file): Rename this to...
(read_genotype_file): ... this.
(genotype_db_put): Put "file_info".
(genotype_db_get): Fetch "file_info".
(import_genotype): Use read_genotype_file.
(print_current_matrix): Remove transpose.  Print out file_info
metadata.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Add module doc-string.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

Alexanderlacuna reviewed (approved):
LGTM!

BonfaceKilz committed:
Delete un-used import.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
More efficiently and correctly count nrows is genotype.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Add functionality for importing one entire genotype directory.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Set a larger map size for the larger genotype files.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Add kludge comment for "HSNIH-Palmer_true.geno" which has floats.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>
